apps/cmp.c

```diff
@@ -0,0 +1,3292 @@
+    }
+
+    if (opt_certsform_s != NULL
+            && !opt_format(opt_certsform_s, OPT_FMT_PEMDER | OPT_FMT_PKCS12,
+                           &opt_certsform)) {
+        CMP_err("unknown option given for certificate list loading format");
+        return 0;
+    }
+
+    return 1;
+}
+
+static OSSL_CMP_SRV_CTX *setup_srv_ctx(ENGINE *e)
+{
+    OSSL_CMP_CTX *ctx = NULL; /* extra CMP (client) ctx partly used by server */
+    OSSL_CMP_SRV_CTX *srv_ctx = ossl_cmp_mock_srv_new();
+
+    if (srv_ctx == NULL)
+        return NULL;
+    ctx = OSSL_CMP_SRV_CTX_get0_cmp_ctx(srv_ctx);
+
+    if (opt_srv_ref == NULL) {
+        if (opt_srv_cert == NULL) {
+            /* opt_srv_cert should determine the sender */
+            CMP_err("must give -srv_ref for server if no -srv_cert given");
+            goto err;
+        }
+    } else {
+        if (!OSSL_CMP_CTX_set1_referenceValue(ctx, (unsigned char *)opt_srv_ref,
+                                              strlen(opt_srv_ref)))
+            goto err;
+    }
+
+    if (opt_srv_secret != NULL) {
+        int res;
+        char *pass_str = get_passwd(opt_srv_secret, "PBMAC secret of server");
+
+        if (pass_str != NULL) {
+            cleanse(opt_srv_secret);
+            res = OSSL_CMP_CTX_set1_secretValue(ctx, (unsigned char *)pass_str,
+                                                strlen(pass_str));
+            clear_free(pass_str);
+            if (res == 0)
+                goto err;
+        }
+    } else if (opt_srv_cert == NULL) {
+        CMP_err("server credentials must be given if -use_mock_srv or -port is used");
+        goto err;
+    } else {
+        CMP_warn("server will not be able to handle PBM-protected requests since -srv_secret is not given");
+    }
+
+    if (opt_srv_secret == NULL
+            && ((opt_srv_cert == NULL) != (opt_srv_key == NULL))) {
+        CMP_err("must give both -srv_cert and -srv_key options or neither");
+        goto err;
+    }
+    if (opt_srv_cert != NULL) {
+        X509 *srv_cert = STORE_load_cert(opt_srv_cert, opt_srv_keypass,
+                                         "certificate of the server");
+        /* from server perspective the server is the client */
+        if (srv_cert == NULL || !OSSL_CMP_CTX_set1_clCert(ctx, srv_cert)) {
+            X509_free(srv_cert);
+            goto err;
+        }
+        X509_free(srv_cert);
+    }
+    if (opt_srv_key != NULL) {
+        EVP_PKEY *pkey = load_key(opt_srv_key, opt_keyform, 0, opt_srv_keypass,
+                                  e, "private key for server cert");
+
+        if (pkey == NULL || !OSSL_CMP_CTX_set1_pkey(ctx, pkey)) {
+            EVP_PKEY_free(pkey);
+            goto err;
+        }
+        EVP_PKEY_free(pkey);
+    }
+    cleanse(opt_srv_keypass);
+
+    if (opt_srv_trusted != NULL) {
+        X509_STORE *ts =
+            load_certstore(opt_srv_trusted, "certificates trusted by server");
+        char *NULL_host = NULL; /* for CMP level, no host etc */
```

Comment:
perhaps call this `null_host`? Why do we even need this? We never seem to write to it.

Reply:
This was a leftover (at two points); removed